nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py

```diff
@@ -0,0 +1,317 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import itertools
+import math
+import random
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import numpy as np
+import torch
+import torch.utils.data as pt_data
+from hydra.utils import instantiate
+from omegaconf import DictConfig
+from pytorch_lightning import Trainer
+from pytorch_lightning.utilities import rank_zero_only
+from sacrebleu import corpus_bleu
+
+from nemo.collections.common.losses import SmoothedCrossEntropyLoss
+from nemo.collections.common.metrics import Perplexity
+from nemo.collections.common.parts import transformer_weights_init
+from nemo.collections.nlp.data import TranslationDataset
+from nemo.collections.nlp.data.machine_translation.machine_translation_dataset import TranslationDataConfig
+from nemo.collections.nlp.models.enc_dec_nlp_model import EncDecNLPModel, EncDecNLPModelConfig
+from nemo.collections.nlp.modules.common.transformer import BeamSearchSequenceGenerator
+from nemo.core.classes.common import typecheck
+from nemo.utils import logging
+
+__all__ = ['MTEncDecModel']
+
+
+@dataclass
+class MTEncDecModelConfig(EncDecNLPModelConfig):
+    train_ds: Optional[TranslationDataConfig] = None
+    validation_ds: Optional[TranslationDataConfig] = None
+    test_ds: Optional[TranslationDataConfig] = None
+    num_val_examples: int = 3
+    num_test_examples: int = 3
+    beam_size: int = 1
+    len_pen: float = 0.0
+    max_generation_delta: int = 50
+    label_smoothing: Optional[float] = 0.0
+    vocab_divisible_by_eight: bool = True
+
+
+class MTEncDecModel(EncDecNLPModel):
+    """
+    Left-to-right Transformer language model.
+    """
+
+    def __init__(self, cfg: MTEncDecModelConfig, trainer: Trainer = None):
+
+        self.setup_enc_dec_tokenizers(cfg)
+
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        # make vocabulary size divisible by 8 for fast fp16 training
+        if cfg.vocab_divisible_by_eight:
+            self.encoder_vocab_size = 8 * math.ceil(self.encoder_tokenizer.vocab_size / 8)
```

Comment:
this will not have effect because tokenizers didn't actually change.
To simplify things, how about for now, we drop ```cfg.vocab_divisible_by_eight``` all together?

Reply:
(so, for now, it won't autopad embedding matrices to be divisible by 8).

Reply:
That's fine with me.